rfcs/experimental/kernel_templates/README.md

```diff
@@ -0,0 +1,294 @@
+
+## Proposal
+
+### High-Level Structure
+
+The algorithms are defined in `<oneapi/dpl/experimental/kernel_templates>`,
+in `namespace oneapi::dpl::experimental::kt`.
+This namespace contains portable algorithms and namespaces with more specialized implementations.
+
+Below is an example of such a structure with possible algorithms:
+
+```c++
+namespace oneapi::dpl::experimental::kt
+{
+// Portable algorithms
+/*return*/ reduce(/*args*/);
+/*return*/ transform(/*args*/);
+/*return*/ inclusive_scan(/*args*/);
+/*return*/ radix_sort(/*args*/);
+// ...
+
+  namespace gpu
+  {
+    // Algorithms optimized for GPU architectures
+    /*return*/ reduce(/*args*/);
+    /*return*/ transform(/*args*/);
+    // ...
+
+    namespace esimd
+    {
+      // Algorithms optimized for Intel GPUs supporting ESIMD technology.
+      /*return*/ reduce(/*args*/);
+      /*return*/ transform(/*args*/);
+      // ...
+    }
+  } // namespace oneapi::dpl::experimental::kt::gpu
+
+  namespace cpu
+  {
+    // Algorithms optimized for CPU architectures
+    // ...
+  }
+
+} // namespace oneapi::dpl::experimental::kt
+```
+
+There is no requirement for the sets of algorithms in each namespace to be identical.
+
+### Abstract Algorithm Signature
+
+A kernel template is a C++ function invoked from the host.
+
+Each function takes a `sycl::queue` as its first argument and
+an instance of the [kernel_param](kernel_configuration/README.md) template,
+which specializes parameters common to all kernel templates, as its last argument.
+The arguments in between vary depending on the specific algorithm.
+The function may also have algorithm-specific non-deducible template parameters.
+
+Below is an abstract signature of a kernel template:
+
+```c++
+// defined in <oneapi/dpl/experimental/kernel_templates>
+// in namespace oneapi::dpl::experimental::kt
+
+template <
+    typename AlgorithmParameter1,  // manually specified, may have a default
+    typename AlgorithmParameter2,  // ...
+    typename AlgorithmParameterN,  // manually specified, may have a default
+
+    typename KernelParam,          // deduced, may have a default in future
+
+    typename Arg1,                 // deduced, may have a default
+    typename Arg2,                 // ...
+    typename ArgN                  // deduced, may have a default
+>
+sycl::event kernel_template (
+    sycl::queue q,
+    /* any cvref */ Arg1 arg1,
+    /* any cvref */ Arg2 arg2,
+    /* any cvref */ ArgN argn,
+    KernelParam param              // manually specified, may have a default in future
+);
+```
+
+`Arg1`, ..., `ArgN` include sequences to process.
+Data must be passed using the same mechanisms described in the
+[documentation on passing data](https://uxlfoundation.github.io/oneDPL/parallel_api/pass_data_algorithms.html#pass-data-to-algorithms)
+for algorithms with device policies.
+Specialized algorithms may extend or restrict the supported data passing mechanisms.
```

Comment:
Do we intend to support host allocated data in Kernel templates?
  
I'd suggest we require data to be indirectly device accessible, given that these are intended to favor performance over productivity and this is sycl specific.  However, if we want to support buffers with `begin/end` (which I think we do), we will still need to process the input, so it may not make much of a technical difference in ease of implementation.  We may want to detect and exclude "bad" data, and / or process it a little differently from mainline oneDPL algorithms.